docs(nx-cloud): add new self-healing ci feature page (#31851)
nrwl/nx: docs/shared/features/self-healing-ci.md

---
title: 'AI-Powered Self-Healing CI'
description: 'Learn how Nx Cloud Self-Healing CI uses AI to automatically detect, analyze, and fix CI failures, eliminating the need to babysit PRs and keeping you focused on building features.'
keywords: [self-healing CI, AI, CI automation]
---

# Self-Healing CI

{% youtube
src="https://youtu.be/JW5Ki3PkRWA"
title="Introducing Self-Healing CI for Nx and Nx Cloud"
/%}

Nx Cloud Self-Healing CI is an **AI-powered system that automatically detects, analyzes, and proposes fixes for CI failures**, offering several key advantages:

- **Improves Time to Green (TTG):** Significantly reduces the time it takes to get your PR merge-ready by automatically proposing fixes when your tasks fail. No more baby-sitting PRs until they are ready for review.

> - **Keeps You in the Flow:** With Nx Console, you get direct notifications in your editor (VS Code, Cursor, or WebStorm) showing failed PRs and proposed fixes. Review, approve, and keep working while the AI handles the rest in the background.

Reviewer: ```suggestion
- **Keeps You in the Flow:** With Nx Console, you get direct notifications in your editor (VS Code, Cursor, or JetBrain IDEs) showing failed PRs and proposed fixes. Review, approve, and keep working while the AI handles the rest in the background.
```

## Enable Self-Healing CI

To enable Self-Healing CI in your workspace, you'll need to connect to Nx Cloud and configure your CI pipeline.

### 1. Connect to Nx Cloud

If you haven't already connected to Nx Cloud:

```shell
npx nx connect
```

If you don't have Nx Cloud yet, you can [start with the free Hobby plan](/pricing) and explore the AI features.

Make sure AI features are enabled in your [Nx Cloud dashboard](https://nx.app):

1. **[Organization Settings](https://cloud.nx.app/go/organization/edit)**: Enable AI features (required for all AI-powered functionality)
2. **Workspace Settings**: Find the "Self-Healing CI" section and make sure it is toggled on

### 2. Configure your CI pipeline

#### Option A: Using Nx Agents (Recommended)

If you're using [Nx Agents](/ci/features/distribute-task-execution), Self-Healing CI is automatically enabled:

```yaml {% fileName=".github/workflows/ci.yml" highlightLines=[12] %}
name: CI

jobs:
  main:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
        with:
          fetch-depth: 0
          filter: tree:0

>       - run: npx nx-cloud start-ci-run --distribute-on="3 linux-medium-js" --stop-agents-after="build"

Reviewer: ```suggestion
      - run: npx nx-cloud start-ci-run --distribute-on="3 linux-medium-js" --stop-agents-after="build,test,lint"
```

          cache: 'npm'

      - run: npm ci
      - uses: nrwl/nx-set-shas@v4
      - run: npx nx run-many -t lint test build
```

#### Option B: Without Nx Agents

If you're not using Nx Agents, add the `fix-ci` step to your pipeline. **Important**: This step must run at the end with `if: always()` to ensure it executes even when previous steps fail:

```yaml {% fileName=".github/workflows/ci.yml" highlightLines=[15,16] %}
name: CI

jobs:
  main:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
        with:
          fetch-depth: 0

      - run: npm ci
      - uses: nrwl/nx-set-shas@v4
      - run: npx nx affected -t lint test build

      - run: npx nx cloud fix-ci
        if: always()
```

### 3. Install Nx Console

To receive notifications about self-healing activities directly in your editor, install [Nx Console](/getting-started/editor-setup) for VS Code, Cursor, or IntelliJ.

For the complete AI setup guide, see our [AI integration documentation](/getting-started/ai-integration).

## How Self-Healing CI works

Here's what happens when you push a PR with Self-Healing CI enabled:

1. **You push your PR** - Nothing changes in your workflow.

> 2. **Failure detected** - If tasks fail, instead of just reporting the failure, Nx Cloud starts an AI agent.

Reviewer: ```suggestion
2. **Failure detected** - If tasks fail, instead of reporting the failure, Nx Cloud starts an AI agent.
```

3. **AI agent analyzes** - The agent examines the error logs, understands your codebase structure through Nx's project graph, and identifies the root cause.

> 4. **Fix proposed** - The agent creates a fix and presents it to you via Nx Console directly in your editor (VS Code, Cursor, or WebStorm) or through the integrated GitHub application.

Reviewer: ```suggestion
4. **Fix proposed** - The agent creates a fix and presents it to you via Nx Console directly in your editor (VS Code, Cursor, or JetBrain IDEs) or through the integrated GitHub application.
```

small nit, not limited to only webstorm

5. **Validation runs in parallel** - Meanwhile, the agent validates the fix by re-running the originally failed tasks with the proposed changes. This step is optional and can be skipped but serves as an additional verification step that the fix works.

> 6. **Human review and approval** - You get a notification in your editor showing the failed PR and proposed fix in an integrated view. You can review, validate, and approve the fix without leaving your development environment.

Reviewer: ```suggestion
6. **Human review and approval** - You get a notification in your editor showing the failed PR and proposed fix. You can review, validate, and approve the fix without leaving your development environment.
```

> 7. **Automatic PR update** - Once you approve, the fix gets committed to your PR as a new commit by the AI agent while you continue working on other tasks.

Reviewer: ```suggestion
7. **Automatic PR update** - Once approve, the fix is applied as a new commit to the PR by the AI agent while you continue working on other tasks.
```

> 8. **Full CI re-run** - Your complete CI pipeline runs again with the applied fix.

Reviewer: ```suggestion
8. **Full CI re-run** - A new CI Pipeline execution is started with the applied fix.
```
